### PR TITLE
[lldb] Skip remote PutFile when MD5 hashes equal

### PR DESCRIPTION
**Pull request by Awfa**
This PR adds a check within `PutFile` to exit early when both local and destination files have matching MD5 hashes. If they differ, or there is trouble getting the hashes, the regular code path to put the file is run.

As I needed this to talk to an `lldb-server` which runs the gdb-remote protocol, I enabled `CalculateMD5` within `Platform/gdb-server` and also found and fixed a parsing bug within it as well. Before this PR, the client is incorrectly parsing the response packet containing the checksum; after this PR, hopefully this is fixed. There is a test for the parsing behavior included in this PR.

### CODE REVIEW

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp:689]: Suggestion: Flip the condition and do an early return with false.

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp:690]: It's good that you have a comment for this because it "feels" wrong. But why are `high/low` swapped in the gdb remote communication client in the first place? Is that intentional or is this working around a bug?

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3440]: Why not fix the bug instead of working around it here? What causes the bug?

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3450]: Can you add some meaning to `sizeof(uint64_t) * 2`? It's used a lot, so giving it a name would help with the readability.

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3456]: suggestion: `if (part.getAsInteger(/*radix=*/16, low))`

File: lldb/source/Target/Platform.cpp
Comment (bulbazord) [inline lldb/source/Target/Platform.cpp:1187]: Why did this get changed?

File: lldb/source/Target/Platform.cpp
Comment (bulbazord) [inline lldb/source/Target/Platform.cpp:1202]: Why is this all in its own block?

File: lldb/source/Target/Platform.cpp
Comment (bulbazord) [inline lldb/source/Target/Platform.cpp:1213]: LLDB uses c++17 so you can use structured binding for this.
```
const auto [local_md5_high, local_md5_low] = local_md5->words();
```

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (Awfa) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3440]: `response.GetHexMaxU64` consumes the response stream until a non-valid hex character is encountered.

This is just the wrong function to use because the server returns both the low/high parts concatenated. When it was used, what would happen is `low = response.GetHexMaxU64...` would consume the whole packet, when it should just be consuming the first half of it. This also leads to `high` being set to 0 because `low` consumed the entire packet.

This part of the patch isn't a workaround, it's the fix. This fixes the client so it can correctly parse the server's response.

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (Awfa) [reply lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp:689]: Should I change the other functions in `PlatformRemoteGDBServer` to do the same? For example `PlatformRemoteGDBServer::GetFileExists` does the same currently.

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (Awfa) [reply lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp:690]: Good point - I don't see a reason why it is swapped there. I'll change it so it's just consistent.

File: lldb/source/Target/Platform.cpp
Comment (Awfa) [reply lldb/source/Target/Platform.cpp:1187]: When testing this change, and using `log enable lldb platform` to read the logs, I just noticed this had a `\n` and removed it since `LLDB_LOGF` already adds a new line, and it didn't seem like most other logs had `\n`.

Is this bad to have in the patch?

File: lldb/source/Target/Platform.cpp
Comment (Awfa) [reply lldb/source/Target/Platform.cpp:1202]: This is how I write code to scope the variables `dest_md5_low`, `dest_md5_high`, `success` away from the main scope of the function.

Is this not good to do in the llvm-project?

File: lldb/source/Target/Platform.cpp
Comment (Awfa) [reply lldb/source/Target/Platform.cpp:1213]: Will do

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (Awfa) [reply lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp]: I'll change just this instance to meet your suggestion

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (DavidSpickett) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3440]: Maybe a small wording change then "which is a bug" could be "which would give incorrect results".

"is a bug" sounds a lot like we have a bug still.

File: lldb/source/Target/Platform.cpp
Comment (DavidSpickett) [reply lldb/source/Target/Platform.cpp:1187]: Since (I assume) you can't push directly, I suggest leaving this out of this PR and once this PR lands immediately put up a PR just to change that \n.

It's maybe a bit strict but PRs are free :)

File: lldb/source/Target/Platform.cpp
Comment (DavidSpickett) [reply lldb/source/Target/Platform.cpp:1202]: If you really need a new scope, a small static function in the same file is the usual way. In this case it's fine to have them in the scope of the main function I think, I don't think they overlap with any existing var.

**Comment by DavidSpickett** [comment]
@weliveindetail I think this might fix the problems you were having remote debugging clang-repl (I can't seem to find the actual Discourse thread).

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (bulbazord) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3440]: I agree with David here. My initial impression is that you were doing things this way to work around a bug in `GetHexMaxU64`.

File: lldb/source/Target/Platform.cpp
Comment (bulbazord) [reply lldb/source/Target/Platform.cpp:1187]: I wouldn't say it's bad to have it, but I wasn't sure why it was here. It seems unrelated to the intent of this PR, which is why I asked about it. I'm fine with you leaving it in this PR or submitting this in a different PR.

File: lldb/source/Target/Platform.cpp
Comment (bulbazord) [reply lldb/source/Target/Platform.cpp:1202]: I see. I wouldn't say it's a bad thing to create a new scope if you want to contain the lifetime of these variables to their own scope, but in LLDB (and more broadly LLVM) this is mostly used for RAII. The most common instance of this is when one wants to acquire a lock but does not want the lock to span the lifetime of the method body.

For consistency I would say either put this code in the main body, put it in a small static function, or put it in a lambda.

File: lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp
Comment (bulbazord) [inline lldb/source/Plugins/Platform/gdb-server/PlatformRemoteGDBServer.cpp:691]: The LLVM coding standards suggest not having curly braces for 1 line bodies: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by bulbazord — COMMENTED** [review]
Looking better! Let's get the comments and style sorted out, but I think this is looking good :)

File: lldb/source/Target/Platform.cpp
Comment (Awfa) [reply lldb/source/Target/Platform.cpp]: I decided to leave it out of this PR

File: lldb/source/Target/Platform.cpp
Comment (Awfa) [reply lldb/source/Target/Platform.cpp]: Just removed the scoping - so the md5 vars are part of the main scope.

**Comment by emaste** [comment]
MD5 is insufficient for claiming that files are identical; how do we migrate this to a secure hash?

**Comment by Awfa** [comment]
> MD5 is insufficient for claiming that files are identical; how do we migrate this to a secure hash?

Is there an attack vector you're concerned about?

Or are you wary of workflow friction when a file won't upload to the remote platform because the hashes accidently collide?

**Comment by emaste** [comment]
> > MD5 is insufficient for claiming that files are identical; how do we migrate this to a secure hash?
> 
> Is there an attack vector you're concerned about?
> 
> Or are you wary of workflow friction when a file won't upload to the remote platform because the hashes accidently collide?

It's the latter I'm thinking of. I don't imagine this particularly interesting or valuable to an attacker, but MD5 collisions are becoming quite easy to obtain and I think it's plausible some software will make use of intentional collisions. Mainly I think we should have a plan to move to a better hash -- is there any interop issue with GDB's protocol/stub?

**Comment by DavidSpickett** [comment]
If it were a GDB packet it would be in https://sourceware.org/gdb/current/onlinedocs/gdb.html/Host-I_002fO-Packets.html#Host-I_002fO-Packets but I see no sign of it, or in GDB's sources.

The first appearance of `vFile:MD5` is e0f8f574c7f41f5b61ec01aa290d52fc55a3f3c9 though it is not documented as one of our extensions in https://github.com/llvm/llvm-project/blob/main/lldb/docs/lldb-platform-packets.txt.

So a good follow up PR would be to list it in that document. It's self explanatory but still, weird that it's not there.

This means we could add `vFile:betterHash` (or a more general `hash:<kind>`) and try sending it to the remote. If the remote responds that it doesn't support it, we ask it for MD5. Though we may want a way to say don't use MD5 even if the remote would accept it, if security is the concern (a bit like ssh key formats work).

In the case of this specific PR, it's fixing code that should have worked all along but didn't. So the issue of MD5 collision is worth looking at, but I wouldn't let it block this.

@emaste, since you know about the shortcomings of MD5, could you open an issue for this? And I'll add to that the internal lldb details.

**Comment by DavidSpickett** [comment]
@bulbazord please give the final approval if it looks good to you.

We can sort out moving from MD5 on an issue.

**Review by bulbazord — APPROVED** [review]
LGTM

**Comment by emaste** [comment]
Ok, submitted https://github.com/llvm/llvm-project/issues/89271 for the MD5 migration. I agree that issue does not block this change.

**Comment by Awfa** [comment]
Someone will have to merge for me because it says "Only those with write access to this repository can merge pull requests."

Am I supposed to request write access somewhere before making these PRs? I haven't contributed before.

**Comment by bulbazord** [comment]
> Someone will have to merge for me because it says "Only those with write access to this repository can merge pull requests."
> 
> Am I supposed to request write access somewhere before making these PRs? I haven't contributed before.

First time contributors aren't expected to have commit access, so no worries there. I approved it so I can land this change for you. Let's keep an eye on the build bots in case this inadvertently introduces any regressions. :)

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:3422]: If we're changing the interface anyway, could we have this return an `std::optional<llvm::MD5::MD5Result>` to match what `md5_contents` returns?

File: lldb/source/Target/Platform.cpp
Comment (JDevlieghere) [inline lldb/source/Target/Platform.cpp:1207]: It's not obvious from the right hand side what the return value is. Case in point, it returns an `ErrorOr<MD5::MD5Result>` and if I hadn't looked it up I wouldn't have noticed we're not actually logging the error.

**Comment by Awfa** [comment]
I got an email that the build failed and I should revert because of this: https://lab.llvm.org/buildbot/#/builders/68/builds/72623

@bulbazord can you or someone with write permissions revert this PR so I have time to triage the issue?

**Comment by DavidSpickett** [comment]
> So a good follow up PR would be to list it in that document. It's self explanatory but still, weird that it's not there.

https://github.com/llvm/llvm-project/pull/89357

**Comment by DavidSpickett** [comment]
> @bulbazord can you or someone with write permissions revert this PR so I have time to triage the issue?

The next build is green, those DAP tests do fail once in a while so it's unrelated to your change.

**Comment by DavidSpickett** [comment]
Please address @JDevlieghere's comments in a new PR.

**Comment by bulbazord** [comment]
> I got an email that the build failed and I should revert because of this: https://lab.llvm.org/buildbot/#/builders/68/builds/72623
> 
> @bulbazord can you or someone with write permissions revert this PR so I have time to triage the issue?

As David said, no need since that was unrelated to your change. Jonas makes a good point in his review though (that I totally missed when I reviewed), please open a new PR that addresses his feedback.

**Comment by Awfa** [comment]
Thanks for taking a look at the build. I will do a new PR addressing the issues.

**Pull request merged**
